### PR TITLE
[TableGen] Optimize intrinsic info type signature encoding

### PR DESCRIPTION
**Pull request by jurahul**
Change the "fixed encoding" table used for encoding intrinsic
type signature to use 16-bit encoding as opposed to 32-bit.

This results in both space and time improvements. For space,
the total static storage size (in bytes) of this info reduces by 50%:
- Current = 14193*4 (Fixed table) + 16058 + 3 (Long Table) = 72833
- New size = 14193*2 (Fixed table) + 19879 + 3 (Long Table) = 48268.
- Reduction = 50.9%

For time, with the added benchmark, we see a 7.3% speedup in 
`GetIntrinsicInfoTableEntries` benchmark. Actual output of the 
benchmark in included in the GitHub MR.

### CODE REVIEW

**Comment by jurahul** [comment]
Here's the actual benchmark run results:

```
32-bit$ ./build/benchmarks/DecodeIITFixedEncodingBM --benchmark_repetitions=20 --benchmark_display_aggregates_only
2024-09-01T07:27:27-07:00
Running ./build/benchmarks/DecodeIITFixedEncodingBM
Run on (20 X 2918.4 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 24576 KiB (x1)
Load Average: 0.38, 0.16, 0.12
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_DecodeIITFixedEncoding_mean              97373 ns        97372 ns           20
BM_DecodeIITFixedEncoding_median            97194 ns        97193 ns           20
BM_DecodeIITFixedEncoding_stddev              881 ns          881 ns           20
BM_DecodeIITFixedEncoding_cv                 0.90 %          0.90 %            20
BM_GetIntrinsicInfoTableEntries_mean       543853 ns       543848 ns           20
BM_GetIntrinsicInfoTableEntries_median     542532 ns       542528 ns           20
BM_GetIntrinsicInfoTableEntries_stddev       4759 ns         4758 ns           20
BM_GetIntrinsicInfoTableEntries_cv           0.88 %          0.87 %            20

16-bit$ ./build/benchmarks/DecodeIITFixedEncodingBM --benchmark_repetitions=20 --benchmark_display_aggregates_only
2024-09-01T07:28:23-07:00
Running ./build/benchmarks/DecodeIITFixedEncodingBM
Run on (20 X 2918.4 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 24576 KiB (x1)
Load Average: 0.43, 0.22, 0.14
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_DecodeIITFixedEncoding_mean              43832 ns        43832 ns           20
BM_DecodeIITFixedEncoding_median            43675 ns        43675 ns           20
BM_DecodeIITFixedEncoding_stddev              632 ns          632 ns           20
BM_DecodeIITFixedEncoding_cv                 1.44 %          1.44 %            20
BM_GetIntrinsicInfoTableEntries_mean       475647 ns       475647 ns           20
BM_GetIntrinsicInfoTableEntries_median     474756 ns       474755 ns           20
BM_GetIntrinsicInfoTableEntries_stddev       5090 ns         5090 ns           20
BM_GetIntrinsicInfoTableEntries_cv           1.07 %          1.07 %            20

```
% speedups:
* DecodeIITFixedEncoding  : 97372 -> 43832 = 55%
* GetIntrinsicInfoTableEntries : 543848 -> 475647 = 12.5%

**Comment by jurahul** [comment]
Actual benchmark runs:

```
Old:

~/upstream_clean$ ./build/benchmarks/GetIntrinsicInfoTableEntriesBM --benchmark_repetitions=40 --benchmark_display_aggregates_only
2024-09-01T08:16:40-07:00
Running ./build/benchmarks/GetIntrinsicInfoTableEntriesBM
Run on (20 X 2918.4 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 24576 KiB (x1)
Load Average: 0.43, 0.29, 0.40
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_GetIntrinsicInfoTableEntries_mean       511908 ns       511877 ns           40
BM_GetIntrinsicInfoTableEntries_median     510461 ns       510461 ns           40
BM_GetIntrinsicInfoTableEntries_stddev       9411 ns         9428 ns           40
BM_GetIntrinsicInfoTableEntries_cv           1.84 %          1.84 %            40


new:

~/upstream_llvm/llvm-project$ ./build/benchmarks/GetIntrinsicInfoTableEntriesBM --benchmark_repetitions=40 --benchmark_display_aggregates_only
2024-09-01T08:15:59-07:00
Running ./build/benchmarks/GetIntrinsicInfoTableEntriesBM
Run on (20 X 2918.4 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 24576 KiB (x1)
Load Average: 0.19, 0.23, 0.39
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_GetIntrinsicInfoTableEntries_mean       474298 ns       474293 ns           40
BM_GetIntrinsicInfoTableEntries_median     474271 ns       474270 ns           40
BM_GetIntrinsicInfoTableEntries_stddev       5285 ns         5287 ns           40
BM_GetIntrinsicInfoTableEntries_cv           1.11 %          1.11 %            40
```

Speedup = 511877  -> 474293  = 7.3%

File: llvm/utils/TableGen/IntrinsicEmitter.cpp [hunk 1]
```diff
@@ -282,11 +283,34 @@ static TypeSigTy ComputeTypeSignature(const CodeGenIntrinsic &Int) {
...
+  }
+  return Result;
+}
+
 void IntrinsicEmitter::EmitGenerator(const CodeGenIntrinsicTable &Ints,
                                      raw_ostream &OS) {
-  // If we can compute a 32-bit fixed encoding for this intrinsic, do so and
+  // Note: the code below can be switched to use 32-bit fixed encoding by
+  // flipping the flag below.
+  constexpr bool Use16BitFixedEncoding = true;
```
Comment (jurahul) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:304]: Not sure if we want to keep this flexibility around. I guess it's easy enough to go back to 32-bit in future if needed, so this flexibility not needed, and I can simplify this. WDYT?

File: llvm/benchmarks/GetIntrinsicInfoTableEntriesBM.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "benchmark/benchmark.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/Intrinsics.h"
+#include <variant>
```
Comment (arsenm) [inline llvm/benchmarks/GetIntrinsicInfoTableEntriesBM.cpp:12]: I don't see variant used here

File: llvm/utils/TableGen/IntrinsicEmitter.cpp [hunk 3]
```diff
@@ -282,11 +282,34 @@ static TypeSigTy ComputeTypeSignature(const CodeGenIntrinsic &Int) {
...
+  using EncodingTy =
+      std::conditional_t<Use16BitFixedEncoding, uint16_t, unsigned>;
+  const unsigned Mask = Use16BitFixedEncoding ? 0x7FFF : 0x7FFFFFFF;
+  const unsigned MSBPostion = Use16BitFixedEncoding ? 15 : 31;
```
Comment (arsenm) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:307]: Make these computed based on the size of the type?

File: llvm/utils/TableGen/IntrinsicEmitter.cpp [hunk 4]
```diff
@@ -282,11 +282,34 @@ static TypeSigTy ComputeTypeSignature(const CodeGenIntrinsic &Int) {
...
+  StringRef TypeName = Use16BitFixedEncoding ? "uint16_t" : "unsigned";
```
Comment (arsenm) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:308]: Consistently use uintN_t

File: llvm/utils/TableGen/IntrinsicEmitter.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/utils/TableGen/IntrinsicEmitter.cpp:303]: I think the cost of keeping it around is low and reduces the friction for further experimentation

Comment (jurahul) [reply llvm/utils/TableGen/IntrinsicEmitter.cpp:303]: SG, I'll keep it.

File: llvm/benchmarks/GetIntrinsicInfoTableEntriesBM.cpp (hunk 2 above)
Comment (jurahul) [reply llvm/benchmarks/GetIntrinsicInfoTableEntriesBM.cpp]: Leftover from an earlier version. Removed now.

File: llvm/utils/TableGen/IntrinsicEmitter.cpp (hunk 3 above)
Comment (jurahul) [reply llvm/utils/TableGen/IntrinsicEmitter.cpp]: Done.

File: llvm/utils/TableGen/IntrinsicEmitter.cpp [hunk 5]
```diff
@@ -282,11 +282,36 @@ static TypeSigTy ComputeTypeSignature(const CodeGenIntrinsic &Int) {
   return TypeSig;
 }
 
+// Pack the type signature into 32-bit fixed encoding word.
+std::optional<uint32_t> encodePacked(const TypeSigTy &TypeSig) {
```
Comment (arsenm) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:286]: static

**Review by arsenm — APPROVED** [review]

File: llvm/utils/TableGen/IntrinsicEmitter.cpp (hunk 5 above)
Comment (jurahul) [reply llvm/utils/TableGen/IntrinsicEmitter.cpp]: Done

**Pull request merged**